python/src/iceberg/avro/decoder.py

```diff
@@ -0,0 +1,193 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import datetime
+import decimal
+import struct
+
+from iceberg.io.base import InputStream
+
+STRUCT_FLOAT = struct.Struct("<f")  # little-endian float
+STRUCT_DOUBLE = struct.Struct("<d")  # little-endian double
+STRUCT_SIGNED_SHORT = struct.Struct(">h")  # big-endian signed short
+STRUCT_SIGNED_INT = struct.Struct(">i")  # big-endian signed int
+STRUCT_SIGNED_LONG = struct.Struct(">q")  # big-endian signed long
+
+
+class BinaryDecoder:
+    """Read leaf values."""
+
+    _input_stream: InputStream
+
+    def __init__(self, input_stream: InputStream) -> None:
+        """
+        reader is a Python object on which we can call read, seek, and tell.
+        """
+        self._input_stream = input_stream
+
+    def read(self, n: int) -> bytes:
+        """
+        Read n bytes.
+        """
+        if n < 0:
+            raise ValueError(f"Requested {n} bytes to read, expected positive integer.")
+        read_bytes = self._input_stream.read(n)
+        if len(read_bytes) != n:
```

Comment:
Is this a guarantee from `InputStream`? I thought that it was possible to return fewer bytes.

Reply:
I don't think that's possible. We know how many bytes we need to read, if we got fewer that indicates that something is off.

Reply:
Actually, I think that we can have this happen. `InputStream` is a protocol to describe our requirements for IO classes, which is intended to allow passing something that extends `IOBase`. The [docs for `RawIOBase`](https://docs.python.org/3/library/io.html#io.RawIOBase) show that there isn't a guarantee that all the bytes will be returned:

> Fewer than size bytes may be returned if the operating system call returns fewer than size bytes.

I think we should handle small reads. Maybe add a `read_fully` method since we always want to get the exact number of bytes that are expected?

Reply:
Alright, I tried to implement this which isn't that hard, but we also have other interfaces that might not have implemented the `read_fully`, breaking the protocol of InputStream. This happens with PyArrows NativeFile as well:
```python
    def test_pyarrow_input_file():
        """Test reading a file using PyArrowFile"""
    
        with tempfile.TemporaryDirectory() as tmpdirname:
            file_location = os.path.join(tmpdirname, "foo.txt")
            with open(file_location, "wb") as f:
                f.write(b"foo")
    
            # Confirm that the file initially exists
            assert os.path.exists(file_location)
    
            # Instantiate the input file
            absolute_file_location = os.path.abspath(file_location)
            input_file = PyArrowFile(location=f"{absolute_file_location}")
    
            # Test opening and reading the file
            f = input_file.open()
>           assert isinstance(f, InputStream)  # Test that the file object abides by the InputStream protocol
E           assert False
E            +  where False = isinstance(<pyarrow.lib.NativeFile object at 0x111bbf6f0>, InputStream)
```

Looks like pyarrow does return the exact number of bytes, and also has a `read1` method fwiw:

<img width="1530" alt="image" src="https://user-images.githubusercontent.com/1134248/173676133-4f8aac53-9d49-4c93-bcb7-d32a3cd52c93.png">

Reply:
I might not understand you correctly, but it sounds like you're adding `read_fully` to the Protocol and that's why PyArrow's `NativeFile` doesn't conform to `InputStream`? I was thinking that we should add an IO util class that has a `read_fully` method, like we have in Java: https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/io/IOUtil.java

Reply:
Then we should add another class in between. Adding it to the protocol means that it should be implemented to NativeFile as well. Happy to do this, but we can also do this in a separate PR.

Reply:
Yeah, I agree. We should have some additional methods that help fill the gap between the IO API and what we need for usage.